### PR TITLE
React Locators in Puppeteer

### PR DESCRIPTION
**Pull request by DavertMik**
Added React locators to Puppeteer

Also, I had to update the documentation generation process to include React instructions for every method that support react locators (without changing the text in other helpers). 

* Add React locators support to Puppeteer
* Added tests for React locators for WebDriver and Puppeteer
* Removed Gulp
* Moved docs generation to RoboFile.php
* Updated documentation.js

### CODE REVIEW

File: docs/helpers/Protractor.md [hunk 1]
```diff
@@ -326,88 +322,123 @@ Opposite to `see`. Checks that a text is not present on a page.
...
+
 #### Parameters
 
--   `field`  located by label|name|CSS|XPath|strict locator.
+-   `field` **([string][9] \| [object][10])** located by label|name|CSS|XPath|strict locator.
 
 ### dontSeeCookie
 
-Checks that cookie with given name does not exist.
+{{> dontSeeCookie}}
```
[the remark this answers is not in the document]
Comment (LukoyanovE) [reply docs/helpers/Protractor.md]: or this a type of link, which not visible in simple markdown?

[the remark this answers is not in the document]
Comment (DavertMik) [reply docs/helpers/Protractor.md]: thanks, you are right, templating issue

File: docs/helpers/Appium.md [hunk 2]
```diff
@@ -135,7 +139,7 @@ I.click('Submit');
 // CSS button
 I.click('#form input[type=submit]');
 // XPath
-I.click('//form/*[@type=submit]');
```
Comment (LukoyanovE) [inline docs/helpers/Appium.md]: Looks like a `*` should not be removed

File: docs/helpers/Appium.md [hunk 3]
```diff
@@ -197,8 +205,9 @@ I.dontSeeElement('.modal'); // modal is not shown
 
 #### Parameters
 
--   `locator` **([string][4] \| [object][5])** located by CSS|XPath|Strict locator.
--   `context`
```
Comment (LukoyanovE) [inline docs/helpers/Appium.md]: `context` argument is lost

File: docs/helpers/Nightmare.md [hunk 4]
```diff
@@ -399,43 +440,53 @@ I.fillField({css: 'form#login input[name=username]'}, 'John');
...
 
--   `name`  
+-   `name`  cookie name. 
+
+Returns [Promise][8]&lt;[string][3]> attribute value
+
+
+
+Returns any Promise<object> Cookie in JSON format. If name not passed returns all cookies for this domain.Multiple cookies can be received by passing query object:```js
+I.grabCookie({ secure: true});
```
Comment (LukoyanovE) [inline docs/helpers/Nightmare.md]: Either multiple string or without `js`.

`js` is prepend string:
It looks like:```js I.grabCookie({ secure: true}); ```asdasd 

When multiple string:
```js
I.grabCookie({ secure: true});
```

File: docs/helpers/Nightmare.md [hunk 5]
```diff
@@ -807,6 +922,10 @@ I.seeNumberOfElements('#submitBtn', 1);
 
 -   `selector`  
 -   `num`  
+-   `locator` ([string][3] \| [object][4]) element located by CSS|XPath|strict locator.
+-   `number` [number][9] of elements.
```
Comment (LukoyanovE) [inline docs/helpers/Nightmare.md]: There are duplication of differently names arguments again.

File: docs/helpers/Protractor.md [hunk 6]
```diff
@@ -23,15 +23,15 @@ This helper should be configured in codecept.json or codecept.conf.js
...
--   `smartWait`: (optional) **enables [SmartWait][3]**; wait for additional milliseconds for element to appear. Enable for 5 secs: "smartWait": 5000
+-   `smartWait`: (optional) enables [SmartWait][3]; wait for additional milliseconds for element to appear. Enable for 5 secs: "smartWait": 5000
 -   `disableScreenshots`   - don't save screenshot on failure
 -   `fullPageScreenshots`  - make full page screenshots on failure.
 -   `uniqueScreenshotNames`   - option to prevent screenshot override if you have scenarios with the same name in different suites
 -   `keepBrowserState`   - keep browser state between tests when `restart` set to false.
 -   `seleniumAddress` - Selenium address to connect (default: [http://localhost:4444/wd/hub][4])
 -   `rootElement` - Root element of AngularJS application (default: body)
 -   `getPageTimeout` (optional) sets default timeout for a page to be loaded. 10000 by default.
--   `waitForTimeout`: (optional) sets default wait time in _ms_ for all `wait*` functions. 1000 by default.
```
Comment (LukoyanovE) [inline docs/helpers/Protractor.md]: Should `*`  be removed?

File: docs/helpers/Protractor.md [hunk 7]
```diff
@@ -1358,10 +1560,12 @@ I.waitNumberOfVisibleElements('a', 3);
...
 
--   `locator` **([string][9] \| [object][10])** element located by CSS|XPath|strict locator.
+-   `locator` ([string][9] \| [object][10]) element located by CSS|XPath|strict locator.
 -   `num`  
--   `sec` **[number][7]** (optional, `1` by default) time in seconds to wait 
--   `number` **[number][7]** of elements.
+-   `sec` [number][7] (optional, `1` by default) time in seconds to wait
+    
+ 
+-   `number` [number][7] of elements.
```
Comment (LukoyanovE) [inline docs/helpers/Protractor.md]: Arguments duplication

File: docs/helpers/Puppeteer.md [hunk 8]
```diff
@@ -30,7 +30,7 @@ This helper should be configured in codecept.json or codecept.conf.js
 -   `waitForAction`: (optional) how long to wait after click, doubleClick or PressKey actions in ms. Default: 100.
 -   `waitForNavigation`: . When to consider navigation succeeded. Possible options: `load`, `domcontentloaded`, `networkidle0`, `networkidle2`. See [Puppeteer API][2]. Array values are accepted as well.
 -   `getPageTimeout`  config option to set maximum navigation time in milliseconds.
--   `waitForTimeout`: (optional) default wait\* timeout in ms. Default: 1000.
+-   `waitForTimeout`: (optional) default wait\ timeout in ms. Default: 1000.
```
Comment (LukoyanovE) [inline docs/helpers/Puppeteer.md]: Should `*` be removed?  (3)

File: docs/helpers/SeleniumWebdriver.md [hunk 9]
```diff
@@ -36,13 +36,13 @@ This helper should be configured in codecept.json or codecept.conf.js
 -   `browser` - browser in which perform testing
 -   `driver` - which protractor driver to use (local, direct, session, hosted, sauce, browserstack). By default set to 'hosted' which requires selenium server to be started.
 -   `restart` - restart browser between tests (default: true).
--   `smartWait`: (optional) **enables SmartWait**; wait for additional milliseconds for element to appear. Enable for 5 secs: "smartWait": 5000
+-   `smartWait`: (optional) enables SmartWait; wait for additional milliseconds for element to appear. Enable for 5 secs: "smartWait": 5000
 -   `disableScreenshots`   - don't save screenshot on failure
 -   `uniqueScreenshotNames`   - option to prevent screenshot override if you have scenarios with the same name in different suites
 -   `keepBrowserState`   - keep browser state between tests when `restart` set to false.
--   `keepCookies`   - keep cookies between tests when `restart` set to false.\*
```
Comment (LukoyanovE) [inline docs/helpers/SeleniumWebdriver.md]: Should `*` be removed ?(4)

File: docs/helpers/SeleniumWebdriver.md [hunk 10]
```diff
@@ -36,13 +36,13 @@ This helper should be configured in codecept.json or codecept.conf.js
...
+-   `keepCookies`   - keep cookies between tests when `restart` set to false.\
 -   `seleniumAddress` - Selenium address to connect (default: [http://localhost:4444/wd/hub][7])
--   `waitForTimeout`: (optional) sets default wait time in _ms_ for all `wait*` functions. 1000 by default;
```
Comment (LukoyanovE) [inline docs/helpers/SeleniumWebdriver.md]: Should `*` be removed ?(5)

File: docs/helpers/WebDriver.md [hunk 11]
```diff
@@ -1319,9 +1451,11 @@ I.seeNumberOfElements('#submitBtn', 1);
 
 #### Parameters
 
--   `locator` **([string][18] \| [object][19])** element located by CSS|XPath|strict locator.
+-   `locator` ([string][18] \| [object][19]) element located by CSS|XPath|strict locator.
 -   `num`  
--   `number` **[number][21]** of elements.-   -   _Appium_: supported
+-   `number` [number][21] of elements.
```
Comment (LukoyanovE) [inline docs/helpers/WebDriver.md]: Still arguements duplication

File: docs/helpers/WebDriver.md [hunk 12]
```diff
@@ -1334,8 +1468,15 @@ I.seeNumberOfVisibleElements('.buttons', 3);
...
--   `locator` **([string][18] \| [object][19])** element located by CSS|XPath|strict locator.
--   `num` **[number][21]** number of elements.-   {{react}}
+-   `locator` ([string][18] \| [object][19]) element located by CSS|XPath|strict locator.
+-   `num` [number][21] number of elements.
+    
+
+    
+
+
+This action supports [React locators](https://codecept.io/react#locators)
```
Comment (LukoyanovE) [inline docs/helpers/WebDriver.md]: Hm, 6 line breaks?

File: lib/helper/Appium.js [hunk 13]
```diff
@@ -1256,9 +1256,8 @@ class Appium extends Webdriver {
 
   /**
    * {{> dontSeeElement }}
-   *
    */
-  async dontSeeElement(locator, context) {
```
Comment (LukoyanovE) [inline lib/helper/Appium.js]: Why `context` is removed?

File: docs/helpers/Protractor.md [hunk 14]
```diff
@@ -595,34 +695,41 @@ let postHTML = await I.grabHTMLFrom('#post');
...
-Retrieves a page scroll position and returns it to test.
-Resumes test execution, so **should be used inside an async function with `await`** operator.
 
-```js
-let { x, y } = await I.grabPageScrollPosition();
-```
+
+### grabPageScrollPosition
+
+{{> grabPageScrollPosition}}
```
Comment (LukoyanovE) [inline docs/helpers/Protractor.md]: Template still not fixed

File: docs/helpers/Puppeteer.md [hunk 15]
```diff
@@ -658,33 +807,18 @@ let url = await I.grabCurrentUrl();
...
-{ // all results are in [ms]
-responseEnd: 23,
-domInteractive: 44,
-domContentLoadedEventEnd: 196,
-loadEventEnd: 241
-}
-```
+### grabDataFromPerformanceTiming
+
+{{> ../webapi/grabDataFromPerformanceTiming }}
```
Comment (LukoyanovE) [inline docs/helpers/Puppeteer.md]: Template error

File: docs/helpers/Puppeteer.md [hunk 16]
```diff
@@ -695,34 +829,46 @@ let postHTML = await I.grabHTMLFrom('#post');
...
-```
+
+
+
+This action supports [React locators](https://codecept.io/react#locators)
+
+
+### grabPageScrollPosition
+
+{{> grabPageScrollPosition}}
```
Comment (LukoyanovE) [inline docs/helpers/Puppeteer.md]: Template error

File: docs/helpers/WebDriver.md [hunk 17]
```diff
@@ -899,47 +1035,51 @@ If multiple elements found returns an array of texts.
...
-Extra shift can be set with offsetX and offsetY options.
 
-```js
-I.moveCursorTo('.tooltip');
-I.moveCursorTo('#submit', 5,5);
-```
+
+### moveCursorTo
+
+{{> moveCursorTo}}
```
Comment (LukoyanovE) [inline docs/helpers/WebDriver.md]: Template error

File: docs/helpers/WebDriver.md [hunk 18]
```diff
@@ -963,36 +1103,10 @@ I.pressKey(['Control','a']);
...
-    -   `'Pause'`,
-    -   `'Semicolon'`,
-    -   `'Shift'`,
-    -   `'Space'`,
-    -   `'Subtract'`,
-    -   `'Tab'`.To make combinations with modifier and mouse clicks (like Ctrl+Click) press a modifier, click, then release it.-   _Appium_: supported, but clear field before pressing in apps:```js
+-   `key` ([string][18] \| [array][26]) key or array of keys to press.
+    
+
+    {{> \_keys }}To make combinations with modifier and mouse clicks (like Ctrl+Click) press a modifier, click, then release it.```js
```
Comment (LukoyanovE) [inline docs/helpers/WebDriver.md]: template error

File: docs/helpers/WebDriver.md [hunk 19]
```diff
@@ -1063,19 +1188,12 @@ Placeholder for ~ locator only test case write once run on both Appium and WebDr
...
 
-Saves a screenshot to ouput folder (set in codecept.json or codecept.conf.js).
-Filename is relative to output folder. 
-Optionally resize the window to the full available page `scrollHeight` and `scrollWidth` to capture the entire page by passing `true` in as the second argument.
-
-```js
-I.saveScreenshot('debug.png');
-I.saveScreenshot('debug.png', true) //resizes to available scrollHeight and scrollWidth before taking screenshot
-```
+{{> saveScreenshot}}
```
Comment (LukoyanovE) [inline docs/helpers/WebDriver.md]: template error

File: lib/helper/Appium.js (hunk 13 above)
Comment (DavertMik) [reply lib/helper/Appium.js]: It was not there

File: docs/helpers/SeleniumWebdriver.md (hunk 10 above)
Comment (DavertMik) [reply docs/helpers/SeleniumWebdriver.md]: this helper is deprecated, not even shown on website :)

File: docs/helpers/Protractor.md [hunk 20]
```diff
@@ -1227,7 +1236,10 @@ I.seeNumberOfElements('#submitBtn', 1);
 #### Parameters
 
 -   `selector`  
--   `num`  
+-   `num` [number][7] number of elements.
+    
+
+-   `locator` ([string][9] \| [object][10]) element located by CSS|XPath|strict locator.
```
Comment (LukoyanovE) [inline docs/helpers/Protractor.md]: Arguments duplication again

File: docs/helpers/Puppeteer.md [hunk 21]
```diff
@@ -1364,23 +1407,24 @@ I.seeInTitle('Home Page');
...
 
-This action supports [React locators](https://codecept.io/react#locators)
-
 
-#### Parameters
+This action supports [React locators](https://codecept.io/react#locators)
 
--   `selector`  
--   `num`  
+-   `locator` ([string][7] \| [object][5]) element located by CSS|XPath|strict locator.
```
Comment (LukoyanovE) [inline docs/helpers/Puppeteer.md]: Arguments duplication again

File: docs/helpers/WebDriver.md [hunk 22]
```diff
@@ -1452,10 +1502,10 @@ I.seeNumberOfElements('#submitBtn', 1);
 #### Parameters
 
 -   `locator` ([string][18] \| [object][19]) element located by CSS|XPath|strict locator.
--   `num`  
--   `number` [number][21] of elements.
+-   `num` [number][21] number of elements.
     
 
+    {{ react}}
```
Comment (LukoyanovE) [inline docs/helpers/WebDriver.md]: Template error

File: docs/helpers/Nightmare.md [hunk 23]
```diff
@@ -921,11 +919,10 @@ I.seeNumberOfElements('#submitBtn', 1);
 #### Parameters
 
 -   `selector`  
--   `num`  
--   `locator` ([string][3] \| [object][4]) element located by CSS|XPath|strict locator.
--   `number` [number][9] of elements.
+-   `num` [number][9] number of elements.
     
 
+-   `locator` ([string][3] \| [object][4]) element located by CSS|XPath|strict locator.
```
Comment (LukoyanovE) [inline docs/helpers/Nightmare.md]: Arguments duplication

**Pull request merged**